<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Defining your own types · JuliaForBioinformatics</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>JuliaForBioinformatics</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li><span class="toctext">Introduction</span><ul><li><a class="toctext" href="../01_Introduction/">Introduction</a></li><li><a class="toctext" href="../02_DataStructures/">Data Structures</a></li><li><a class="toctext" href="../03_Files_Homework/">HOMEWORK: Working with Files</a></li></ul></li><li><span class="toctext">Advance</span><ul><li><a class="toctext" href="../04_HowJuliaWorks/">How Julia works?</a></li><li class="current"><a class="toctext" href>Defining your own types</a><ul class="internal"></ul></li><li><a class="toctext" href="../06_Strings/">Working with Strings</a></li><li><a class="toctext" href="../07_Stats/">Statistics</a></li><li><a class="toctext" href="../08_Packages_Homework/">HOMEWORK: Packages</a></li></ul></li><li><span class="toctext">Bioinformatics</span><ul><li><a class="toctext" href="../09_MSA/">Multiple Sequence Alignments</a></li><li><a class="toctext" href="../10_Information/">Counting residues</a></li><li><a class="toctext" href="../11_PDB/">MSA and structures</a></li></ul></li><li><a class="toctext" href="../api/">API</a></li></ul></nav><article id="docs"><header><nav><ul><li>Advance</li><li><a href>Defining your own types</a></li></ul><a class="edit-page" href="https://github.com/diegozea/JuliaForBioinformatics/blob/master/src/Advance/05_DefiningTypes.jl"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Defining your own types</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Defining-your-own-types-1" href="#Defining-your-own-types-1">Defining your own types</a></h1><p><a href="https://mybinder.org/v2/gh/diegozea/JuliaForBioinformatics/gh-pages?filepath=dev/notebooks/05_DefiningTypes.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt/></a> <a href="https://nbviewer.jupyter.org/github/diegozea/JuliaForBioinformatics/blob/gh-pages/dev/notebooks/05_DefiningTypes.ipynb"><img src="https://img.shields.io/badge/show-nbviewer-579ACA.svg" alt/></a></p><p>Julia user-defined types are as fast as built-in types. They are easy to define:</p><pre><code class="language-julia">abstract type AbstractFeature end</code></pre><pre><code class="language-julia">struct Exon{T} &lt;: AbstractFeature where {T &lt;: Integer}
	start::T
	stop::T
end</code></pre><pre><code class="language-julia">Exon(5, 10)</code></pre><pre><code class="language-none">Main.ex-05_DefiningTypes.Exon{Int64}(5, 10)</code></pre><pre><code class="language-julia">mutable struct MessengerRNA
	sequence::Vector{Char}
	exons::Vector{Exon}
	present::Vector{Bool}
end</code></pre><pre><code class="language-julia">mRNA = MessengerRNA(collect(&quot;ACTGTTGCATTGCAATTTAAGCAATGGCAAATAACATA&quot;),
					[Exon(5,10), Exon(20,28), Exon(30,34)],
					[true, false, true])</code></pre><pre><code class="language-none">Main.ex-05_DefiningTypes.MessengerRNA([&#39;A&#39;, &#39;C&#39;, &#39;T&#39;, &#39;G&#39;, &#39;T&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;A&#39;, &#39;T&#39;  …  &#39;A&#39;, &#39;A&#39;, &#39;A&#39;, &#39;T&#39;, &#39;A&#39;, &#39;A&#39;, &#39;C&#39;, &#39;A&#39;, &#39;T&#39;, &#39;A&#39;], Main.ex-05_DefiningTypes.Exon[Exon{Int64}(5, 10), Exon{Int64}(20, 28), Exon{Int64}(30, 34)], Bool[true, false, true])</code></pre><p>You can access type fields using a dot:</p><pre><code class="language-julia">mRNA.exons</code></pre><pre><code class="language-none">3-element Array{Main.ex-05_DefiningTypes.Exon,1}:
 Main.ex-05_DefiningTypes.Exon{Int64}(5, 10) 
 Main.ex-05_DefiningTypes.Exon{Int64}(20, 28)
 Main.ex-05_DefiningTypes.Exon{Int64}(30, 34)</code></pre><h4><a class="nav-anchor" id="Exercise-1-1" href="#Exercise-1-1">Exercise 1</a></h4><p>Define a function that takes a MessengerRNA object and return a list with the sequences (as strings) of the exons present in the transcript, i.e. <code>[&quot;TTGCAT&quot;, &quot;AATAA&quot;]</code></p><pre><code class="language-julia"># ... your function here ...</code></pre><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p><footer><hr/><a class="previous" href="../04_HowJuliaWorks/"><span class="direction">Previous</span><span class="title">How Julia works?</span></a><a class="next" href="../06_Strings/"><span class="direction">Next</span><span class="title">Working with Strings</span></a></footer></article></body></html>
